syntax = "proto3";

package todo;

// Todo message definition
message Todo {
  string id = 1;
  string title = 2;
  string description = 3;
  string priority = 4;
  bool completed = 5;
  string due_date = 6;
  string user_id = 7;
  string created_at = 8;
  string updated_at = 9;
}

// Request messages
message CreateTodoRequest {
  string title = 1;
  string description = 2;
  string priority = 3;
  string due_date = 4;
  string user_id = 5;
}

message GetTodoRequest {
  string id = 1;
}

message UpdateTodoRequest {
  string id = 1;
  string title = 2;
  string description = 3;
  string priority = 4;
  bool completed = 5;
  string due_date = 6;
}

message DeleteTodoRequest {
  string id = 1;
}

message ListTodosRequest {
  int32 page = 1;
  int32 limit = 2;
  string user_id = 3;
}

// Response messages
message TodoResponse {
  Todo todo = 1;
}

message ListTodosResponse {
  repeated Todo todos = 1;
  int32 total = 2;
}

message DeleteTodoResponse {
  bool success = 1;
}

// gRPC service definition
service TodoService {
  rpc CreateTodo(CreateTodoRequest) returns (TodoResponse);
  rpc GetTodo(GetTodoRequest) returns (TodoResponse);
  rpc UpdateTodo(UpdateTodoRequest) returns (TodoResponse);
  rpc DeleteTodo(DeleteTodoRequest) returns (DeleteTodoResponse);
  rpc ListTodos(ListTodosRequest) returns (ListTodosResponse);
}
